### PR TITLE
Fix ParticleFilter::resample()

### LINKED ISSUE #232: ParticleFilter::resample() does not remove first particle collectly
Screenshot from 2019-06-14 15-58-09

In the screenshot above, there is a particle far away from others. This wrong particle did not disappear after the robot moved.
This happens because there is a bug in resampling algorithm.
In ParticleFilter::resample(), the first particle always remains because the value of "pscan" is always 0 when i = 0.

### PR DESCRIPTION
**Pull request by nhatao**
Closes #232 

* Fix resample() to remove first particle correctly.
* A test is added to verify this fix.

### CODE REVIEW

**Comment by at-wat** [comment]
## [#599] PASSED on kinetic

<details><summary>All tests passed</summary>


```
build/test_results/mcl_3dl/gtest-test_chunked_kdtree.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_filter.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_imu_measurement_model_gravity.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_motion_prediction_model_differential_drive.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_nd.xml: 2 tests
build/test_results/mcl_3dl/gtest-test_pf.xml: 6 tests
build/test_results/mcl_3dl/gtest-test_point_cloud_random_sampler.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_point_types.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_quat.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_raycast.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_vec3.xml: 5 tests
build/test_results/mcl_3dl/roslint-mcl_3dl.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_expansion_resetting_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_global_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_mcl_3dl_compat_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_tf_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_transform_rostest.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_pose.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_tf.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_expansion_resetting.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_global_localization.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_mcl_3dl_compat.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_transform_failure.xml: 1 tests
Summary: 44 tests, 0 errors, 0 failures, 0 skipped
```
</details>

**Comment by at-wat** [comment]
## [#600] PASSED on kinetic

<details><summary>All tests passed</summary>


```
build/test_results/mcl_3dl/gtest-test_chunked_kdtree.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_filter.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_imu_measurement_model_gravity.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_motion_prediction_model_differential_drive.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_nd.xml: 2 tests
build/test_results/mcl_3dl/gtest-test_pf.xml: 6 tests
build/test_results/mcl_3dl/gtest-test_point_cloud_random_sampler.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_point_types.xml: 1 tests
build/test_results/mcl_3dl/gtest-test_quat.xml: 4 tests
build/test_results/mcl_3dl/gtest-test_raycast.xml: 3 tests
build/test_results/mcl_3dl/gtest-test_vec3.xml: 5 tests
build/test_results/mcl_3dl/roslint-mcl_3dl.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_expansion_resetting_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_global_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_localization_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_mcl_3dl_compat_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_tf_rostest.xml: 1 tests
build/test_results/mcl_3dl/rostest-test_tests_transform_rostest.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_pose.xml: 1 tests
build/test_results/mcl_3dl/rosunit-compare_tf.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_expansion_resetting.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_global_localization.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_mcl_3dl_compat.xml: 1 tests
build/test_results/mcl_3dl/rosunit-test_transform_failure.xml: 1 tests
Summary: 44 tests, 0 errors, 0 failures, 0 skipped
```
</details>

File: include/mcl_3dl/pf.h [hunk 1]
```diff
@@ -180,16 +181,16 @@ class ParticleFilter
     particles_dup_ = particles_;
     std::sort(particles_dup_.begin(), particles_dup_.end());
 
-    const FLT_TYPE pstep = accum / (particles_.size() - 1);
-    FLT_TYPE pscan = 0;
+    const FLT_TYPE pstep = accum / particles_.size();
+    static const FLT_TYPE EPS = std::numeric_limits<FLT_TYPE>::epsilon();
```
Comment (at-wat) [inline include/mcl_3dl/pf.h]: this can be directly used in L192

File: include/mcl_3dl/pf.h [hunk 2]
```diff
@@ -180,16 +181,16 @@ class ParticleFilter
...
     auto it = particles_dup_.begin();
     auto it_prev = particles_dup_.begin();
-
     const FLT_TYPE prob = 1.0 / particles_.size();
-    for (auto& p : particles_)
+    for (size_t i = 0; i < particles_.size(); ++i)
     {
+      auto &p = particles_[i];
+      const FLT_TYPE pscan = pstep * (i + 1) - EPS;
```
Comment (at-wat) [inline include/mcl_3dl/pf.h]:
```suggestion
      const FLT_TYPE pscan = std::nextafter(pstep * (i + 1), static_cast<FLT_TYPE>(0.0));
```
might be better?

File: include/mcl_3dl/pf.h (hunk 1 above)
Comment (nhatao) [reply include/mcl_3dl/pf.h]: Removed by the next change.

File: include/mcl_3dl/pf.h (hunk 2 above)
Comment (nhatao) [reply include/mcl_3dl/pf.h]: Done

**Pull request merged**